### PR TITLE
test(release): Add Python script to test UIs

### PR DESCRIPTION
**Pull request by mmicu**
# Description

Add a Python script that automate the testing of the `Frontend` and `Candid` UIs by using [playwright](https://playwright.dev/python/docs/library). 

# How Has This Been Tested?

The script was executed locally. It should produce a similar output.
Moreover, it fails immediately if it does not find the expected fields as well as warnings and errors.

```
$ python3 scripts/test-uis.py --frontend_url 'http://localhost:4943/?canisterId=bd3sg-teaaa-aaaaa-qaaba-cai' --candid_url 'http://localhost:4943/?canisterId=be2us-64aaa-aaaaa-qaabq-cai&id=bkyz2-fmaaa-aaaaa-qaaaq-cai' --browsers chromium firefox webkit
2023-07-11 20:49:12,779 - INFO - Testing "Frontend UI" at "http://localhost:4943/?canisterId=bd3sg-teaaa-aaaaa-qaaba-cai"
2023-07-11 20:49:13,128 - INFO - Checking "chromium" browser
2023-07-11 20:49:13,869 - INFO - Setting name "my name"
2023-07-11 20:49:13,936 - INFO - Clicking button "Click Me!"
2023-07-11 20:49:14,093 - INFO - "Hello, my name!" found in "#greeting"
2023-07-11 20:49:14,093 - INFO - Checking console logs
2023-07-11 20:49:14,094 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'ambient-light-sensor'."
2023-07-11 20:49:14,095 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'battery'."
2023-07-11 20:49:14,095 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'document-domain'."
2023-07-11 20:49:14,095 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'execution-while-not-rendered'."
2023-07-11 20:49:14,096 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'execution-while-out-of-viewport'."
2023-07-11 20:49:14,096 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'navigation-override'."
2023-07-11 20:49:14,096 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'web-share'."
2023-07-11 20:49:14,097 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'speaker-selection'."
2023-07-11 20:49:14,097 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'conversion-measurement'."
2023-07-11 20:49:14,097 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'focus-without-user-activation'."
2023-07-11 20:49:14,098 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Origin trial controlled feature not enabled: 'interest-cohort'."
2023-07-11 20:49:14,098 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'sync-script'."
2023-07-11 20:49:14,098 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'trust-token-redemption'."
2023-07-11 20:49:14,099 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'vertical-scroll'."
2023-07-11 20:49:14,099 - INFO - Console logs are ok
2023-07-11 20:49:14,209 - INFO - Checking "firefox" browser
2023-07-11 20:49:14,861 - INFO - Setting name "my name"
2023-07-11 20:49:14,909 - INFO - Clicking button "Click Me!"
2023-07-11 20:49:15,041 - INFO - "Hello, my name!" found in "#greeting"
2023-07-11 20:49:15,042 - INFO - Checking console logs
2023-07-11 20:49:15,042 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'ambient-light-sensor'."
2023-07-11 20:49:15,042 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'battery'."
2023-07-11 20:49:15,042 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'document-domain'."
2023-07-11 20:49:15,043 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'execution-while-not-rendered'."
2023-07-11 20:49:15,043 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'execution-while-out-of-viewport'."
2023-07-11 20:49:15,043 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'navigation-override'."
2023-07-11 20:49:15,043 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'web-share'."
2023-07-11 20:49:15,044 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'speaker-selection'."
2023-07-11 20:49:15,044 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'conversion-measurement'."
2023-07-11 20:49:15,044 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'focus-without-user-activation'."
2023-07-11 20:49:15,044 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Origin trial controlled feature not enabled: 'interest-cohort'."
2023-07-11 20:49:15,045 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'sync-script'."
2023-07-11 20:49:15,045 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'trust-token-redemption'."
2023-07-11 20:49:15,045 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'vertical-scroll'."
2023-07-11 20:49:15,045 - INFO - Console logs are ok
2023-07-11 20:49:15,151 - INFO - Checking "webkit" browser
2023-07-11 20:49:15,851 - INFO - Setting name "my name"
2023-07-11 20:49:15,898 - INFO - Clicking button "Click Me!"
2023-07-11 20:49:16,009 - INFO - "Hello, my name!" found in "#greeting"
2023-07-11 20:49:16,009 - INFO - Checking console logs
2023-07-11 20:49:16,010 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'ambient-light-sensor'."
2023-07-11 20:49:16,010 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'battery'."
2023-07-11 20:49:16,010 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'document-domain'."
2023-07-11 20:49:16,010 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'execution-while-not-rendered'."
2023-07-11 20:49:16,010 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'execution-while-out-of-viewport'."
2023-07-11 20:49:16,011 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'navigation-override'."
2023-07-11 20:49:16,011 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'web-share'."
2023-07-11 20:49:16,011 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'speaker-selection'."
2023-07-11 20:49:16,011 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'conversion-measurement'."
2023-07-11 20:49:16,011 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'focus-without-user-activation'."
2023-07-11 20:49:16,012 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Origin trial controlled feature not enabled: 'interest-cohort'."
2023-07-11 20:49:16,012 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'sync-script'."
2023-07-11 20:49:16,012 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'trust-token-redemption'."
2023-07-11 20:49:16,012 - WARNING - Skipping Permissions-Policy warning. log.text="Error with Permissions-Policy header: Unrecognized feature: 'vertical-scroll'."
2023-07-11 20:49:16,013 - INFO - Console logs are ok
2023-07-11 20:49:16,125 - INFO - Testing "Candid UI" at "http://localhost:4943/?canisterId=be2us-64aaa-aaaaa-qaabq-cai&id=bkyz2-fmaaa-aaaaa-qaaaq-cai"
2023-07-11 20:49:16,651 - INFO - Checking "chromium" browser
2023-07-11 20:49:17,555 - INFO - Setting text to "hello, world"
2023-07-11 20:49:17,817 - INFO - Clicking button "Query"
2023-07-11 20:49:17,883 - INFO - Setting text to ""
2023-07-11 20:49:17,912 - INFO - Clicking button "Random"
2023-07-11 20:49:18,011 - INFO - "› greet("hello, world")" found in #output-list at position 0
2023-07-11 20:49:18,011 - INFO - "("Hello, hello, world!")" found in #output-list at position 1
2023-07-11 20:49:18,011 - INFO - "› greet("ti2xjcu")" found in #output-list at position 2
2023-07-11 20:49:18,011 - INFO - "("Hello, ti2xjcu!")" found in #output-list at position 3
2023-07-11 20:49:18,012 - INFO - #output-list lines are defined correctly
2023-07-11 20:49:18,012 - INFO - Checking console logs
2023-07-11 20:49:18,012 - WARNING - Found warning, but it was expected (log.type="Invalid asm.js: Unexpected token", endpoint="/index.js")
2023-07-11 20:49:18,012 - WARNING - Found error, but it was expected (log.type="Failed to load resource: the server responded with a status of 404 (Not Found)", endpoint="/read_state")
2023-07-11 20:49:18,012 - WARNING - Found warning, but it was expected (log.type="Expected to find result for path [object Object], but instead found nothing.", endpoint="/index.js")
2023-07-11 20:49:18,012 - WARNING - Found warning, but it was expected (log.type="Error: Server returned an error:
  Code: 404 (Not Found)
  Body: Custom section name not found.

    at j.readState (http://localhost:4943/index.js:2:11709)
    at async http://localhost:4943/index.js:2:97683
    at async Promise.all (index 0)
    at async Module.UA (http://localhost:4943/index.js:2:98732)
    at async Object.getNames (http://localhost:4943/index.js:2:266156)
    at async http://localhost:4943/index.js:2:275479", endpoint="/index.js")
2023-07-11 20:49:18,012 - INFO - Console logs are ok
2023-07-11 20:49:18,119 - INFO - Checking "firefox" browser
2023-07-11 20:49:18,980 - INFO - Setting text to "hello, world"
2023-07-11 20:49:19,330 - INFO - Clicking button "Query"
2023-07-11 20:49:19,455 - INFO - Setting text to ""
2023-07-11 20:49:19,511 - INFO - Clicking button "Random"
2023-07-11 20:49:19,637 - INFO - "› greet("hello, world")" found in #output-list at position 0
2023-07-11 20:49:19,637 - INFO - "("Hello, hello, world!")" found in #output-list at position 1
2023-07-11 20:49:19,637 - INFO - "› greet("j7usei")" found in #output-list at position 2
2023-07-11 20:49:19,638 - INFO - "("Hello, j7usei!")" found in #output-list at position 3
2023-07-11 20:49:19,638 - INFO - #output-list lines are defined correctly
2023-07-11 20:49:19,638 - INFO - Checking console logs
2023-07-11 20:49:19,639 - WARNING - Found warning, but it was expected (log.type="Invalid asm.js: Unexpected token", endpoint="/index.js")
2023-07-11 20:49:19,639 - WARNING - Found error, but it was expected (log.type="Failed to load resource: the server responded with a status of 404 (Not Found)", endpoint="/read_state")
2023-07-11 20:49:19,639 - WARNING - Found warning, but it was expected (log.type="Expected to find result for path [object Object], but instead found nothing.", endpoint="/index.js")
2023-07-11 20:49:19,640 - WARNING - Found warning, but it was expected (log.type="Error: Server returned an error:
  Code: 404 (Not Found)
  Body: Custom section name not found.

    at j.readState (http://localhost:4943/index.js:2:11709)
    at async http://localhost:4943/index.js:2:97683
    at async Promise.all (index 0)
    at async Module.UA (http://localhost:4943/index.js:2:98732)
    at async Object.getNames (http://localhost:4943/index.js:2:266156)
    at async http://localhost:4943/index.js:2:275479", endpoint="/index.js")
2023-07-11 20:49:19,640 - INFO - Console logs are ok
2023-07-11 20:49:19,729 - INFO - Checking "webkit" browser
2023-07-11 20:49:20,578 - INFO - Setting text to "hello, world"
2023-07-11 20:49:21,112 - INFO - Clicking button "Query"
2023-07-11 20:49:21,221 - INFO - Setting text to ""
2023-07-11 20:49:21,271 - INFO - Clicking button "Random"
2023-07-11 20:49:21,391 - INFO - "› greet("hello, world")" found in #output-list at position 0
2023-07-11 20:49:21,392 - INFO - "("Hello, hello, world!")" found in #output-list at position 1
2023-07-11 20:49:21,392 - INFO - "› greet("ul67tu")" found in #output-list at position 2
2023-07-11 20:49:21,392 - INFO - "("Hello, ul67tu!")" found in #output-list at position 3
2023-07-11 20:49:21,392 - INFO - #output-list lines are defined correctly
2023-07-11 20:49:21,393 - INFO - Checking console logs
2023-07-11 20:49:21,393 - WARNING - Found warning, but it was expected (log.type="Invalid asm.js: Unexpected token", endpoint="/index.js")
2023-07-11 20:49:21,393 - WARNING - Found error, but it was expected (log.type="Failed to load resource: the server responded with a status of 404 (Not Found)", endpoint="/read_state")
2023-07-11 20:49:21,393 - WARNING - Found warning, but it was expected (log.type="Expected to find result for path [object Object], but instead found nothing.", endpoint="/index.js")
2023-07-11 20:49:21,394 - WARNING - Found warning, but it was expected (log.type="Error: Server returned an error:
  Code: 404 (Not Found)
  Body: Custom section name not found.

    at j.readState (http://localhost:4943/index.js:2:11709)
    at async http://localhost:4943/index.js:2:97683
    at async Promise.all (index 0)
    at async Module.UA (http://localhost:4943/index.js:2:98732)
    at async Object.getNames (http://localhost:4943/index.js:2:266156)
    at async http://localhost:4943/index.js:2:275479", endpoint="/index.js")
2023-07-11 20:49:21,394 - INFO - Console logs are ok
2023-07-11 20:49:21,515 - INFO - DONE!
```

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [ ] I have edited the CHANGELOG accordingly.
- [ ] I have made corresponding changes to the documentation.

### CODE REVIEW

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]
Nice!

What we're looking for is a mix between this and the now-closed https://github.com/dfinity/sdk/pull/2519: something we can run as part of a pass/fail e2e test that's run with the rest of our CI checks.  We'll leave the manual verification in scripts/release.sh in place while we build confidence in the e2e test.

It's fine if you want to also run the headless browser test in scripts/release.sh, but it should be pass/fail, rather than involving manual inspection of the output of a python script.  What we're looking for would be cases where the automated test didn't catch something, but the manual verification did.

**Comment by mmicu** [comment]
> Nice!
> 
> What we're looking for is a mix between this and the now-closed #2519: something we can run as part of a pass/fail e2e test that's run with the rest of our CI checks. We'll leave the manual verification in scripts/release.sh in place while we build confidence in the e2e test.

Thanks for the feedback.

So I imagine that I can have something very similar to the closed MR, especially for the YAML configurations, right?
However, I am not sure about the purpose of [.github/workflows/e2e-selenium.yml](https://github.com/dfinity/sdk/pull/2519/files#diff-08832720e778f8eee98325bcafbf82f1abd28fbc606c985e5f33e53b883fd071), since the script is also called from [here](https://github.com/dfinity/sdk/pull/2519/files#diff-3e103440521ada06efd263ae09b259e5507e4b8f7408308dc227621ad9efa31eR191).

> It's fine if you want to also run the headless browser test in scripts/release.sh, but it should be pass/fail, rather than involving manual inspection of the output of a python script. What we're looking for would be cases where the automated test didn't catch something, but the manual verification did.

Then maybe it can be added after the manual inspection, just to confirm that new warnings or new errors are not introduced by a new version.

**Comment by ericswanson-dfinity** [comment]
> However, I am not sure about the purpose of [.github/workflows/e2e-selenium.yml](https://github.com/dfinity/sdk/pull/2519/files#diff-08832720e778f8eee98325bcafbf82f1abd28fbc606c985e5f33e53b883fd071), since the script is also called from [here](https://github.com/dfinity/sdk/pull/2519/files#diff-3e103440521ada06efd263ae09b259e5507e4b8f7408308dc227621ad9efa31eR191).

No need for any equivalent of that e2e-selenium.yml workflow, since we'll be adding one to e2e workflow.

File: .github/workflows/e2e.yml [hunk 1]
```diff
@@ -156,10 +156,50 @@ jobs:
       - name: Run e2e test
         run: timeout 2400 bats "e2e/$E2E_TEST"
 
+  ui_test:
+    runs-on: ${{ matrix.os }}
+    needs: [build_dfx]
+    strategy:
+      matrix:
+        os: [macos-11, ubuntu-20.04]
```
Comment (smallstepman) [inline .github/workflows/e2e.yml:164]: I don't think we need to test across different OSes, right @ericswanson-dfinity ?
```suggestion
        os: [ubuntu-20.04]
```

File: .github/workflows/e2e.yml [hunk 2]
```diff
@@ -156,10 +156,50 @@ jobs:
...
+        rust: ['1.60.0']
+    steps:
+      - name: Checking out repo
+        uses: actions/checkout@v2
```
Comment (smallstepman) [inline .github/workflows/e2e.yml:168]: let's mimic `.jobs.test` more closely
```suggestion
        uses: actions/checkout@v3
```

File: .github/workflows/e2e.yml [hunk 3]
```diff
@@ -156,10 +156,50 @@ jobs:
...
+      - name: Setting up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: '3.9'
+      - name: Installing playwright
+        run: |
+          pip install playwright==1.35.0
+          playwright install
+      - name: Download dfx binary
+        uses: actions/download-artifact@v2
```
Comment (smallstepman) [inline .github/workflows/e2e.yml:178]: let's mimic `.jobs.test` more closely

```suggestion
        uses: actions/download-artifact@v3
```

**Review by smallstepman — CHANGES_REQUESTED** [review]
I've let CI run for this PR, and it bugged out in a bit unexpected place: https://github.com/dfinity/sdk/actions/runs/5591534054/jobs/10227915747?pr=3232

Nothing obvious jumps at me why it crashed there, so I've compared `.jobs.test.steps` and `.jobs.ui_test.steps`; these were some of the differences I was able to find:

File: .github/workflows/e2e.yml (hunk 1 above)
Comment (ericswanson-dfinity) [reply .github/workflows/e2e.yml:164]: We do need to test this essential functionality across different OSes.

**Comment by smallstepman** [comment]
CI is running https://github.com/dfinity/sdk/actions/runs/5602950032?pr=3232

**Comment by smallstepman** [comment]
we're almost there :) https://github.com/dfinity/sdk/actions/runs/5602950032/jobs/10249202379

**Comment by mmicu** [comment]
I found a silly bug in the script. It should solve most of the issues.

However, I cannot make `webkit` work. Is it needed? Locally, `chromium` and `firefox` work fine.

**Comment by smallstepman** [comment]
I'm afraid yes. What is the error message that you're getting?

**Comment by mmicu** [comment]
In the Frontend UI, the `greeting` section is empty once the button is clicked, but it is working for the other two browsers.

I have checked several approaches, but I always got an empty string when I try to retrieve it.

**Comment by mmicu** [comment]
> I'm afraid yes, `webkit` is Safari so it is a major player in the browser game. What is the error message that you're getting?
In the Frontend UI, the `greeting` section is empty once the button is clicked, but it is working for the other two browsers.
I have checked several approaches, but I always got an empty string when I try to retrieve it.

> Btw, have you maybe looked into how hard it would be to also include tests on mobile (android chrome & ios safari)? Having those is basically an endgame for us
Good point. Nope, I did not check it yet.

**Comment by smallstepman** [comment]
What OS are you running this on? I'll check tomorrow on macOS

First thing I'd try to do is try to run launch the browser in [headed mode](https://playwright.dev/python/docs/debug#run-in-debug-mode), have you maybe tried that already?

**Comment by mmicu** [comment]
[That's why it is not working](https://github.com/microsoft/playwright/issues/12975), which cites [[Safari keeps forcing HTTPS on localhost](https://stackoverflow.com/questions/46394682/safari-keeps-forcing-https-on-localhost)](https://stackoverflow.com/questions/46394682/safari-keeps-forcing-https-on-localhost).

I am still checking how to solve it from Playwright.

**Comment by smallstepman** [comment]
Yes, looks like you're right... I tried to look into how to bypass that behavior or how to generate the certificate but didn't found anything so far

Also, when I try to open my locally deployed canister in Safari, the website doesn't load CSS or images. When I try to click on the button, it errors out with `Safari can't open the page because can't establish a secure connection` - @ericswanson-dfinity could you please check if this is also the case for you?

**Comment by mmicu** [comment]
I was able to replicate and fix the issues locally.

Not sure if we actually need to parse the warnings and errors, since we still have the manual checks during the release process.

**Comment by ericswanson-dfinity** [comment]
> @ericswanson-dfinity could you please check if this is also the case for you?

To follow up on this here: this is happening on Safari due to the security policy headers that we introduced in dfx 0.13.1.

File: .github/workflows/e2e.yml [hunk 4]
```diff
@@ -162,10 +162,51 @@ jobs:
       - name: Run e2e test
         run: timeout 2400 bats "e2e/$E2E_TEST"
 
+  ui_test:
+    runs-on: ${{ matrix.os }}
+    needs: [build_dfx]
+    strategy:
+      matrix:
+        os: [macos-12, ubuntu-20.04, ubuntu-22.04]
+        rust: ['1.65.0']
```
Comment (smallstepman) [inline .github/workflows/e2e.yml:171]: we recently bumped rust version in our repo, which is now causing your CI to fail, you'll need to change this 
```suggestion
        rust: ['1.77.1']
```
sorry for the trouble 🥲

Comment (sesi200) [reply .github/workflows/e2e.yml:171]: and if we merge [this PR](https://github.com/dfinity/sdk/pull/3286) it'll be wrong again, too...

Comment (smallstepman) [reply .github/workflows/e2e.yml:171]: why? I don't think it will

File: scripts/test-uis.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+            raise RuntimeError('\n'.join(err))
+
+        # Extract random text from third line
+        random_text = re.search(r'"([^"]*)"', output_list_lines[2])
+        if not random_text:
+            raise RuntimeError(f'Cannot extract the random text from the third line: {output_list_lines[2]}')
+        random_text = random_text.group(1)
+
+        for i, text_str in enumerate([text, random_text]):
+            l1, l2 = (i * 2), (i * 2 + 1)
```
Comment (ericswanson-dfinity) [inline scripts/test-uis.py:208]: `l` and `1` are too close to be used like this in a variable name

```suggestion
            line1, line2 = (i * 2), (i * 2 + 1)
```

File: .github/workflows/e2e.yml [hunk 6]
```diff
@@ -152,10 +152,50 @@ jobs:
...
 
+  ui_test:
+    runs-on: ${{ matrix.os }}
+    needs: [build_dfx]
+    strategy:
+      matrix:
+        os: [macos-12, ubuntu-20.04, ubuntu-22.04]
+    steps:
+      - name: Checking out repo
+        uses: actions/checkout@v3
```
Comment (smallstepman) [inline .github/workflows/e2e.yml:163]:
```suggestion
        uses: actions/checkout@v4
```

**Review by smallstepman — APPROVED** [review]
Amazing work! I'm intrigued by the usage of `_` before each variable/function name, what's the reasoning behind it?

File: .github/workflows/e2e.yml (hunk 6 above)
Comment (mmicu) [reply .github/workflows/e2e.yml:163]: Thanks!

I usually use it to mark objects as private. Probably, in this script they are a bit exaggerated, and we can also consider to remove them.

**Comment by mmicu** [comment]
Hi @ericswanson-dfinity, could you please review the PR? The script seems to work now.

**Review by ericswanson-dfinity — APPROVED** [review]

**Pull request merged**